Unique generic environments

slavapestov opened the pull request.

slavapestov committed:
AST: Replace GenericSignature::createGenericEnvironment() with getGenericEnvironment()

This memoizes the result, which is fine for all callers; the only
exception is open existential types where each new open existential
now explicitly gets a unique generic environment, allocated by
calling GenericEnvironment::getIncomplete().

slavapestov committed:
IRGen: Remove CanGenericSignature::getGenericEnvironment()

slavapestov committed:
AST: Remove GenericEnvironment::OwningDC

slavapestov committed:
Serialization: Simplify GenericEnvironment (de-)serialization

Now that GenericSignatures store their single unique GenericEnvironment,
we can remove similar logic from deserialization to preserve identity
of GenericEnvironments.

slavapestov committed:
ASTVerifier: Simpler archetype verification

jrose-apple commented on include/swift/AST/GenericSignature.h:
```diff
@@ -218,10 +218,13 @@ class alignas(1 << TypeAlignInBits) GenericSignature final
   /// (archetypes) that correspond to the interface types in this generic
   /// signature.
-  GenericEnvironment *createGenericEnvironment();
+  GenericEnvironment *getGenericEnvironment();
```
Should this be `const` (and the field `mutable`)? Although I guess all of GenericSignature is consty, really.

slavapestov replied:
I sort of prefer to use const_cast instead, because then the point of mutation is explicit...

jrose-apple commented:
That one comment regardless, oh wow this feels so much better. Look at that deletion count!

Excited to see if this makes a measurable memory usage improvement.

cc @varungandhi-apple, @brentdax for the Serialization bit

slavapestov commented:
@jrose-apple I'm not sure it will change much wrt memory usage -- you're still getting one generic environment for each non-canonical signature. Also I still need to finish the AST verifier changes. But yeah, the deletions are good.

slavapestov commented:
I did want to try to unique non-canonical generic parameter types by name rather than by generic parameter decl. That should save on memory.

varungandhi-apple commented:
Am I right in understanding that we were not interning the GenericEnvironment before but we are doing that now? Or is there something else going on?

Serialization cleanup looks good.

slavapestov commented:
@varungandhi-apple We were interning them at the serialization level, but @jrose-apple noted that it would be simpler to do it at the AST. This intuition turned out to be correct.

slavapestov committed:
AST: GenericContexts store a GenericSignature instead of a GenericEnvironment

This eliminates the entire 'lazy generic environment' concept;
essentially, all generic environments are now lazy, and since
each signature has exactly one environment, their construction
no longer needs to be co-ordinated with deserialization.

slavapestov committed:
AST: Remove GenericContext::setGenericEnvironment()

This was a transitional step before converting all callers to
store a generic signature instead of an environment.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented:
Oh cool, I didn't expect to see many improvements but it looks like the number of generic signature builders and deserialized conformances decreased somewhat.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test Linux
